framework/src/userobject/DomainUserObject.C

```diff
@@ -0,0 +1,113 @@
+  : UserObject(parameters),
+    BlockRestrictable(this),
+    ThreeMaterialPropertyInterface(this, blockIDs(), Moose::EMPTY_BOUNDARY_IDS),
+    NeighborCoupleable(this, false, false),
+    MooseVariableDependencyInterface(this),
+    TransientInterface(this),
+    RandomInterface(parameters, _fe_problem, _tid, false),
+    ElementIDInterface(this),
+    _mesh(_subproblem.mesh()),
+    _current_elem(_assembly.elem()),
+    _current_elem_volume(_assembly.elemVolume()),
+    _current_side(_assembly.side()),
+    _current_side_elem(_assembly.sideElem()),
+    _current_side_volume(_assembly.sideElemVolume()),
+    _neighbor_elem(_assembly.neighbor()),
+    _current_neighbor_volume(_assembly.neighborVolume()),
+    _current_boundary_id(_assembly.currentBoundaryID()),
+    _normals(_assembly.normals()),
+    _q_point(_assembly.qPoints()),
+    _qrule(_assembly.qRule()),
+    _JxW(_assembly.JxW()),
+    _q_point_face(_assembly.qPointsFace()),
+    _qrule_face(_assembly.qRuleFace()),
+    _JxW_face(_assembly.JxWFace()),
+    _coord(_assembly.coordTransformation())
+{
+  // Keep track of which variables are coupled so we know what we depend on
+  const std::vector<MooseVariableFEBase *> & coupled_vars = getCoupledMooseVars();
+  for (const auto & var : coupled_vars)
+    addMooseVariableDependency(var);
+
+  if (isParamValid("interface_boundaries"))
+  {
+    const auto & interface_boundaries = getParam<std::vector<BoundaryName>>("interface_boundaries");
+    const auto & interface_bnd_ids_vec = _mesh.getBoundaryIDs(interface_boundaries);
+    _interface_bnd_ids =
+        std::set<BoundaryID>(interface_bnd_ids_vec.begin(), interface_bnd_ids_vec.end());
+
+    for (const auto interface_bnd_id : _interface_bnd_ids)
+    {
+      const auto & interface_connected_blocks = _mesh.getInterfaceConnectedBlocks(interface_bnd_id);
+      for (const auto interface_connected_block : interface_connected_blocks)
+      {
+        if (hasBlocks(interface_connected_block))
+          // we're operating on this block
+          continue;
+
+        // these are blocks connected to our blocks
+        _interface_connected_blocks.insert(interface_connected_block);
+      }
+    }
+  }
+}
+
+const MooseVariableFieldBase *
+DomainUserObject::getInterfaceFieldVar(const std::string & var_name, const unsigned int comp)
+{
+  const auto * const field_var = getFieldVar(var_name, comp);
+  mooseAssert(field_var, "We should not be able to return a null variable");
+  _interface_vars.insert(field_var);
+  return field_var;
+}
+
+void
+DomainUserObject::checkVariable(const MooseVariableFieldBase & variable) const
+{
+  if (_interface_vars.count(&variable))
+  {
+    // we could have done this check in the constructor but to be consistent with other block
+    // restrictable checks, we'll do it here
+    for (const auto connected_block : _interface_connected_blocks)
+      if (!variable.hasBlocks(connected_block))
```

Comment:
I guess the loop objects will initialize the variables on interface if the variables are defined on the connecting blocks. My general situation is that a variable may require to be defined on blocks connecting one interface but not all of them. So this check is a bit overzealous or not general enough. We could require an extra parameter in `getInterfaceFieldVar` for the expected interfaces the variable should be defined. Maybe we can skip the check here completely? I am also not sure if `coupledValue` for variable on interfaces will work properly.

Reply:
I would like to do a check if we can. So I would prefer adding another parameter to the method over abandoning all checking.

Why do you think `coupledValue` will not work properly? It binds a reference to `_element_data` in the variable, which is responsible for both volumetric and face variable evaluations. As long as `reinitElemFace` is the last thing called, `coupledValue` will have given us a reference to the correct data

Reply:
I do not recall why I asked the last question, maybe if a variable is not defined on an interface, MOOSE will not initialize it, otherwise we could be in trouble? I like to have the check as well. I will keep working on my changes and see if I can improve the interface here.